bitcoin/src/pow.rs

```diff
@@ -366,6 +366,56 @@ impl CompactTarget {
         Ok(Self::from_consensus(lock_time))
     }
 
+    /// Computes the [`CompactTarget`] from a difficulty adjustment.
+    ///
+    /// ref: <https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp>
+    ///
+    /// Given the previous Target, represented as a [`CompactTarget`], the difficulty is adjusted
+    /// by taking the timespan between them, and multipling the current [`CompactTarget`] by a factor
+    /// of the net timespan and expected timespan. The [`CompactTarget`] may not adjust by more than
+    /// a factor of 4, or adjust beyond the maximum threshold for the network.
+    ///
+    /// # Note
+    ///
+    /// Under the consensus rules, the different in the number of blocks between the headers does
+    /// not equate to the `difficulty_adjustment_interval` of [`Params`]. This is due to an off-by-one
+    /// error, and, the expected number of blocks in between headers is `difficulty_adjustment_interval - 1`
+    /// when calculating the difficulty adjustment.
+    ///
+    /// Take the example of the first difficulty adjustment. Block 2016 introduces a new [`CompactTarget`],
+    /// which takes the net timespan between Block 2015 and Block 0, and recomputes the difficulty.
+    ///
+    /// # Returns
+    ///
+    /// The expected [`CompactTarget`] recalculation.
+    pub fn from_next_work_required(
+        last: CompactTarget,
+        timespan: u64,
+        params: impl AsRef<Params>,
+    ) -> CompactTarget {
+        let params = params.as_ref();
+        if params.no_pow_retargeting {
+            return last;
+        }
+        let mut actual_timespan = timespan;
+        let prev_target: Target = last.into();
+        if timespan.le(&(params.pow_target_timespan / 4)) {
+            actual_timespan = params.pow_target_timespan / 4;
+        }
+        if timespan.ge(&(params.pow_target_timespan * 4)) {
+            actual_timespan = params.pow_target_timespan * 4;
+        }
+        let current_target_int = prev_target.0;
+        let expected_timespan = U256::from(params.pow_target_timespan);
+        let retarget = current_target_int.mul(actual_timespan.into());
+        let retarget = Target(retarget.div(expected_timespan));
+        let maximum_retarget = prev_target.max_transition_threshold(params);
+        if retarget.ge(&maximum_retarget) {
+            return maximum_retarget.to_compact_lossy();
+        }
+        retarget.to_compact_lossy()
```

Comment:
Since this is a direct implementation of the C++ code in Core, and devs are likely to look at both, would it be better to try to mimic the Core code more closely.

Perhaps:
```rust
        // Comments relate to the `pow.cpp` file from Core.

        let maximum_retarget = prev_target.max_transition_threshold(params); // bnPowLimit 

        let retarget = prev_target.0;                             // BnNew
        let retarget = retarget.mul(actual_timespan.into());
        let retarget = retarget.div(params.pow_target_timespan.into());

        let retarget = Target(retarget);

        if retarget.gt(&maximum_retarget) {
            return maximum_retarget.to_compact_lossy();
        }
        retarget.to_compact_lossy()
```